File: Algorithms/Graphs/Kosaraju.js
"use strict";

let setOrder = (graph, vertex, visited, stack) => {

  visited[vertex] = true;

  for (let neighbor = 0; neighbor < graph.length; neighbor++) {

    if (!visited[neighbor] && graph[vertex][neighbor] > 0) {

      setOrder(graph, neighbor, visited, stack);

    }

  }

  stack.push(vertex);

};

let DFS = (graph, vertex, visited, path) => {

  path.push(vertex);
  visited[vertex] = true;

  for (let neighbor = 0; neighbor < graph.length; neighbor++) {

    if (!visited[neighbor] && graph[vertex][neighbor] > 0) {

      DFS(graph, neighbor, visited, path);

    }

  }

};

let transpose = (graph) => graph[0].map((col, i) => graph.map(row => row[i]));

let printSCC = (graph) => {

  let n = graph.length;

  let stack = [],
      visited = (new Array(n)).fill(false);

  for (let vertex = 0; vertex < n; vertex++) {

    if (!visited[vertex]) {

      setOrder(graph, vertex, visited, stack);

    }

  }

  visited = (new Array(n)).fill(false);
  graph = transpose(graph);

  while (stack.length > 0) {

    let vertex = stack.pop();

    if (!visited[vertex]) {

      let path = [];

      DFS(graph, vertex, visited, path);

      console.log(path);

    }

  }

};

let graph = [
  [0, 0, 1, 1, 0],
  [1, 0, 0, 0, 0],
  [0, 1, 0, 0, 0],
  [0, 0, 0, 0, 1],
  [0, 0, 0, 0, 0]
];

printSCC(graph);

File: Algorithms/Graphs/Kosaraju.py

def _set_order(graph, vertex, visited, stack):

    visited[vertex] = True

    for neighbor, dist in enumerate(graph[vertex]):

        if not visited[neighbor] and dist > 0:

            _set_order(graph, neighbor, visited, stack)

    stack.append(vertex)

def _DFS(graph, vertex, visited, path):

    path.append(vertex)
    visited[vertex] = True

    for neighbor, dist in enumerate(graph[vertex]):

        if not visited[neighbor] and dist > 0:

            _DFS(graph, neighbor, visited, path)

def _transpose(graph):

    return list(map(list, zip(*graph)))

def print_SCC(graph):

    n = len(graph)

    stack = []
    visited = [False] * n

    for vertex in range(n):

        if not visited[vertex]:

            _set_order(graph, vertex, visited, stack)

    visited = [False] * n
    graph = _transpose(graph)

    while len(stack) > 0:

        vertex = stack.pop()

        if not visited[vertex]:

            path = []

            _DFS(graph, vertex, visited, path)

            print(path)


def main():

    graph = [
        [0, 0, 1, 1, 0],
        [1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0],
        [0, 0, 0, 0, 1],
        [0, 0, 0, 0, 0]
    ]

    print_SCC(graph)

if __name__ == '__main__':
    main()
